### PR TITLE
feat(Message): added string type for message nonce

### PR DESCRIPTION
**Pull request by wonderlandpark**
**Please describe the changes this PR makes and why it should be merged:**

At [Discord API document](https://discord.com/developers/docs/resources/channel#create-message-params) message `nonce` type is integer or string.
But, for discord.js only accept integer.

**Status**

- [x] Code changes have been tested against the Discord API, or there are no code changes
- [x] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**

- [ ] This PR changes the library's interface (methods or parameters added)
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: src/errors/Messages.js [hunk 1]
```diff
@@ -66,7 +66,7 @@ const Messages = {
   IMAGE_SIZE: size => `Invalid image size: ${size}`,
 
   MESSAGE_BULK_DELETE_TYPE: 'The messages must be an Array, Collection, or number.',
-  MESSAGE_NONCE_TYPE: 'Message nonce must fit in an unsigned 64-bit integer.',
+  MESSAGE_NONCE_TYPE: 'Message nonce must be an Integer or String.',
```
Comment (vladfrangu) [inline src/errors/Messages.js:69]: Slight reword, lmk if you have any better ones 😅

```suggestion
  MESSAGE_NONCE_TYPE: 'Message nonce must be a number or a string.',
```

File: src/structures/APIMessage.js [hunk 2]
```diff
@@ -151,8 +151,8 @@ class APIMessage {
 
     let nonce;
     if (typeof this.options.nonce !== 'undefined') {
-      nonce = parseInt(this.options.nonce);
-      if (isNaN(nonce) || nonce < 0) throw new RangeError('MESSAGE_NONCE_TYPE');
+      nonce = typeof this.options.nonce === 'string' ? this.options.nonce : parseInt(this.options.nonce);
+      if (!nonce) throw new RangeError('MESSAGE_NONCE_TYPE');
```
Comment (vladfrangu) [inline src/structures/APIMessage.js:155]: This check will now not pass if someone sets the nonce to 0... You'll need to adapt the check to account for string or number

File: typings/index.d.ts [hunk 3]
```diff
@@ -999,7 +999,7 @@ declare module 'discord.js' {
     public id: Snowflake;
     public readonly member: GuildMember | null;
     public mentions: MessageMentions;
-    public nonce: string | null;
+    public nonce: string | number | null;
```
Comment (vladfrangu) [inline typings/index.d.ts:1002]: Q: don't you need to update message options too?

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: typings/index.d.ts (hunk 3 above)
Comment (wonderlandpark) [reply typings/index.d.ts:1002]: Agree added for MessageOption and WebhookOption

File: src/errors/Messages.js [hunk 4]
```diff
@@ -66,7 +66,7 @@ const Messages = {
   IMAGE_SIZE: size => `Invalid image size: ${size}`,
 
   MESSAGE_BULK_DELETE_TYPE: 'The messages must be an Array, Collection, or number.',
-  MESSAGE_NONCE_TYPE: 'Message nonce must fit in an unsigned 64-bit integer.',
+  MESSAGE_NONCE_TYPE: 'Message nonce must be a number or a string.',
```
Comment (NotSugden) [inline src/errors/Messages.js:69]: doesn't changing this error message count as a breaking change?

File: src/structures/APIMessage.js [hunk 5]
```diff
@@ -151,8 +151,8 @@ class APIMessage {
 
     let nonce;
     if (typeof this.options.nonce !== 'undefined') {
-      nonce = parseInt(this.options.nonce);
-      if (isNaN(nonce) || nonce < 0) throw new RangeError('MESSAGE_NONCE_TYPE');
+      nonce = typeof this.options.nonce === 'string' ? this.options.nonce : parseInt(this.options.nonce);
+      if (isNaN(nonce) && !nonce) throw new RangeError('MESSAGE_NONCE_TYPE');
```
Comment (NotSugden) [inline src/structures/APIMessage.js:155]: why was the check for a negative nonce removed?
should this check also be a bit more explicit, `typeof nonce === 'number' && ...`?

Comment (kyranet) [inline src/structures/APIMessage.js:155]: If we're going to go strictly for what Discord accepts, the `nonce` field accepts either a string or an integer.

The valid way to check for integer is to use [`Number.isInteger()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isInteger).

The data validation here is very loose, and should be applied only for integers, avoiding further processing on strings.

Please correct me if I'm wrong.

cc: @vladfrangu

Comment (wonderlandpark) [reply src/structures/APIMessage.js:155]: nonce could be negative. It just should be a integer

Comment (wonderlandpark) [reply src/structures/APIMessage.js:155]: The original code was converting param with `parseInt` I also agree check if Number is Integer

Comment (wonderlandpark) [reply src/structures/APIMessage.js:155]: I would like to commit it

Comment (NotSugden) [reply src/structures/APIMessage.js:155]: ah right, what about a check for nonce being `0`, sending the nonce as 0 seems to be ignored by the api

Comment (wonderlandpark) [reply src/structures/APIMessage.js]: Looks like its working

File: src/errors/Messages.js (hunk 4 above)
Comment (wonderlandpark) [reply src/errors/Messages.js:69]: I'm not really sure about it.

File: src/structures/APIMessage.js (hunk 5 above)
Comment (wonderlandpark) [reply src/structures/APIMessage.js]: Made a change for new commit

File: src/structures/APIMessage.js [hunk 6]
```diff
@@ -151,8 +151,8 @@ class APIMessage {
 
     let nonce;
     if (typeof this.options.nonce !== 'undefined') {
-      nonce = parseInt(this.options.nonce);
-      if (isNaN(nonce) || nonce < 0) throw new RangeError('MESSAGE_NONCE_TYPE');
+      nonce = typeof this.options.nonce === 'string' ? this.options.nonce : this.options.nonce;
```
Comment (NotSugden) [inline src/structures/APIMessage.js:154]:
```suggestion
      nonce = this.options.nonce;
```

File: src/structures/APIMessage.js (hunk 5 above)
Comment (NotSugden) [reply src/structures/APIMessage.js]: 0 nonce doesn't seem to be sent back

Comment (wonderlandpark) [reply src/structures/APIMessage.js]: I have checked by sending POST request to direct Discord API and it worked.

Comment (wonderlandpark) [reply src/structures/APIMessage.js]: Please check if i'm wrong!

Comment (vladfrangu) [reply src/structures/APIMessage.js]: I agree with @kyranet here, the check for number should be done only if the nonce IS a number

Comment (wonderlandpark) [reply src/structures/APIMessage.js]: Please check the new commit!

File: src/structures/APIMessage.js [hunk 7]
```diff
@@ -151,8 +151,8 @@ class APIMessage {
 
     let nonce;
     if (typeof this.options.nonce !== 'undefined') {
-      nonce = parseInt(this.options.nonce);
-      if (isNaN(nonce) || nonce < 0) throw new RangeError('MESSAGE_NONCE_TYPE');
+      nonce = this.options.nonce;
+      if ((typeof nonce === 'number' && !Number.isInteger(nonce)) || !nonce) throw new RangeError('MESSAGE_NONCE_TYPE');
```
Comment (SpaceEEC) [inline src/structures/APIMessage.js:155]: This logic does not seem sound to me, some valid values are currently failing here, yet some invalid values pass.

For a false positive example: `0` and `''` fail here, but would pass on Discord's end (although they seem to just drop them)
For a false negative example: `{}` and `[]` pass here but fails on Discord's end.

File: src/errors/Messages.js (hunk 4 above)
Comment (SpaceEEC) [reply src/errors/Messages.js:69]: IMO [`Error#code`](https://github.com/discordjs/discord.js/blob/master/src/errors/DJSError.js#L25) is the semver-relevant part about error messages, but that might be just me being optimistic.
In this case that would be `MESSAGE_NONCE_TYPE`.

Comment (SpaceEEC) [inline src/errors/Messages.js:69]:
```suggestion
  MESSAGE_NONCE_TYPE: 'Message nonce must be a integer or a string.',
```
Floats fail on Discord's end, it's also explicitly documented as `integer`.

**Review by SpaceEEC — CHANGES_REQUESTED** [review]

**Review by kyranet — APPROVED** [review]

File: src/errors/Messages.js [hunk 8]
```diff
@@ -66,7 +66,7 @@ const Messages = {
   IMAGE_SIZE: size => `Invalid image size: ${size}`,
 
   MESSAGE_BULK_DELETE_TYPE: 'The messages must be an Array, Collection, or number.',
-  MESSAGE_NONCE_TYPE: 'Message nonce must fit in an unsigned 64-bit integer.',
+  MESSAGE_NONCE_TYPE: 'Message nonce must be a integer or a string.',
```
Comment (SpaceEEC) [inline src/errors/Messages.js:69]:
```suggestion
  MESSAGE_NONCE_TYPE: 'Message nonce must be an integer or a string.',
```

**Review by SpaceEEC — CHANGES_REQUESTED** [review]
Otherwise LGTM.

**Review by iCrawl — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

**Review by SpaceEEC — APPROVED** [review]

**Comment by wonderlandpark** [comment]
Hello, when will this merged? Is it for the next release?

**Review by iCrawl — APPROVED** [review]

**Pull request merged**